core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -1589,6 +1589,67 @@ private void checkPredicates(RelOptCluster cluster, RelOptTable empTable,
         sortsAs("[IS NULL($0)]"));
   }
 
+  @Test public void testPullUpPredicatesFromUnion() {
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    RelNode rel = null;
+    rel = convertSql(""
+        + "select empno from emp where empno=1\n"
+        + "union all\n"
+        + "select empno from emp where empno=1");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno from emp where empno=1 or deptno=2\n"
+        + "union all\n"
+        + "select empno, deptno from emp where empno=3 or deptno=4");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[OR(=($0, 1), =($1, 2), =($0, 3), =($1, 4))]"));
+
+    rel = convertSql(""
+        + "select empno, comm, deptno from emp where empno=1 and comm=2 and deptno=3\n"
+        + "union all\n"
+        + "select empno, comm, deptno from emp where empno=1 and comm=4");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1), OR(AND(=($1, 2), =($2, 3)), =($1, 4))]"));
+
+  }
+
+  @Test public void testPullUpPredicatesFromIntersect() {
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    RelNode rel = null;
+    rel = convertSql(""
+        + "select empno from emp where empno=1\n"
+        + "intersect all\n"
+        + "select empno from emp where empno=1");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno from emp where empno=1 and deptno=2\n"
+        + "intersect all\n"
+        + "select empno, deptno from emp where empno=1 and deptno=3");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
+
+    rel = convertSql(""
+        + "select empno, deptno, comm from emp where 1=empno and deptno=2\n"
+        + "intersect all\n"
+        + "select empno, deptno, comm from emp where empno=1 and comm=3");
+    assertThat(mq.getPulledUpPredicates(rel).pulledUpPredicates,
+        sortsAs("[=($0, 1)]"));
```

Comment:
Should it be `$0=1 and $1=2 and $3=3`?

Reply:
haisheng ~ you are right ~
My previous thinking was wrong ~
I updated the PR~ The basic idea in current change is to take all pulled up predicates into consideration and remove the redundant ones.